crates/noirc_evaluator/src/ssa_refactor/opt/defunctionalize.rs

```diff
@@ -0,0 +1,369 @@
+//! This module defines the defunctionalization pass for the SSA IR.
+//! The purpose of this pass is to transforms all functions used as values into
+//! constant numbers (fields) that represent the function id. That way all calls
+//! with a non-literal target can be replaced with a call to an apply function.
+//! The apply function is a dispatch function that takes the function id as a parameter
+//! and dispatches to the correct target.
+use std::collections::{HashMap, HashSet};
+
+use acvm::FieldElement;
+use iter_extended::vecmap;
+
+use crate::ssa_refactor::{
+    ir::{
+        basic_block::BasicBlockId,
+        function::{Function, FunctionId, RuntimeType},
+        instruction::{BinaryOp, Instruction},
+        types::{NumericType, Type},
+        value::{Value, ValueId},
+    },
+    ssa_builder::FunctionBuilder,
+    ssa_gen::Ssa,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+struct FunctionSignature {
+    parameters: Vec<Type>,
+    returns: Vec<Type>,
+    runtime: RuntimeType,
+}
+
+impl FunctionSignature {
+    fn from(function: &Function) -> Self {
+        let parameters = vecmap(function.parameters(), |param| function.dfg.type_of_value(*param));
+        let returns = vecmap(function.returns(), |ret| function.dfg.type_of_value(*ret));
+        let runtime = function.runtime();
+        Self { parameters, returns, runtime }
+    }
+}
+
+/// Describes an apply function existing in the SSA
```

Comment:
```suggestion
/// Represents an 'apply' function created by this pass to dispatch higher order functions to.
/// Pseudocode of an `apply` function is given below:
/// ```
/// fn apply(function_id: Field, arg1: Field, arg2: Field) -> Field {
///     match function_id {
///         0 -> function0(arg1, arg2),
///         1 -> function0(arg1, arg2),
///         ...
///         N -> functionN(arg1, arg2),
///     }
/// }
/// ```
/// Apply functions generally take the function to apply as their first parameter. This is a Field value
/// obtained by converting the FunctionId into a Field. The remaining parameters of apply are the
/// arguments to forward to this function when calling it internally.
```